core/server/src/main/java/alluxio/worker/file/UnderFileSystemManager.java

```diff
@@ -0,0 +1,170 @@
+ * Individual streams should only be used by one client. Each instance keeps its own state of
+ * temporary file ids to open streams.
+ */
+@ThreadSafe
+public final class UnderFileSystemManager {
+  /**
+   * A wrapper around the output stream to the under file system. This class handles writing the
+   * data to a temporary file. When the stream is closed, the temporary file will attempt to be
+   * renamed to the final file path. This stream guarantees the temporary file will be cleaned up
+   * when close or cancel is called.
+   */
+  // TODO(calvin): This can be defined by the UnderFileSystem
+  private final class UnderFileSystemOutputStream {
+    /** Configuration to use for this stream. */
+    private final Configuration mConf;
+    /** Underlying stream to the under file system file. */
+    private final OutputStream mStream;
+    /** Final path to write to in the under file system. */
+    private final String mPath;
+    /** Temporary path to write to in the under file system. */
+    private final String mTemporaryPath;
+
+    private UnderFileSystemOutputStream(String ufsPath, Configuration conf)
+        throws FileAlreadyExistsException, IOException {
+      mConf = Preconditions.checkNotNull(conf);
+      mPath = Preconditions.checkNotNull(ufsPath);
+      mTemporaryPath = PathUtils.temporaryFileName(IdUtils.getRandomNonNegativeLong(), mPath);
+      UnderFileSystem ufs = UnderFileSystem.get(mPath, mConf);
+      if (ufs.exists(mPath)) {
+        throw new FileAlreadyExistsException(ExceptionMessage.FAILED_UFS_CREATE.getMessage(mPath));
+      }
+      mStream = ufs.create(mTemporaryPath);
+    }
+
+    /**
+     * Closes the temporary file and deletes it. This object should be discarded after this call.
+     *
+     * @throws IOException if an error occurs during the under file system operation
+     */
+    private void cancel() throws IOException {
+      mStream.close();
+      UnderFileSystem ufs = UnderFileSystem.get(mPath, mConf);
+      // TODO(calvin): Log a warning if the delete fails
+      ufs.delete(mTemporaryPath, false);
+    }
+
+    /**
+     * Closes the temporary file and attempts to promote it to the final file path. If the final
+     * path already exists, the stream is canceled instead.
+     *
+     * @throws IOException if an error occurs during the under file system operation
+     */
+    private void complete() throws IOException {
+      mStream.close();
+      UnderFileSystem ufs = UnderFileSystem.get(mPath, mConf);
+      if (!ufs.rename(mTemporaryPath, mPath)) {
+        // TODO(calvin): Log a warning if the delete fails
+        ufs.delete(mTemporaryPath, false);
+      }
+    }
+
+    /**
+     * @return the wrapped output stream to the under file system file
+     */
+    private OutputStream getStream() {
+      return mStream;
+    }
+  }
+
+  /** A random id generator for worker file ids. */
+  private AtomicLong mIdGenerator;
```

Comment:
If a client tries to talk to a different worker I think it should be a client side bug. If the workers restart and there happens to be a collision between ids, it would be a bug. Although these two cases would both cause unintended behavior, I think the probability is extremely small (similar to how UUIDs could theoretically collide but are considered unique).